scio-avro/src/main/scala/com/spotify/scio/avro/AvroIO.scala

```diff
@@ -175,6 +213,32 @@ final case class SpecificRecordIO[T <: SpecificRecord: ClassTag: Coder](path: St
     SpecificRecordTap[T](ScioUtil.addPartSuffix(path))
 }
 
+final case class SpecificRecordReadFilesIO[T <: SpecificRecord: ClassTag: Coder](
+  paths: Iterable[String]
```

Comment:
I'm a little hesitant about the use of the ReadAll APIs (`FileIO.matchAll()` -> `FileIO.readMatches() -> beam.AvroIO.readFiles()`) here for any `Iterable` of paths. From what I've seen in the documentation/ad hoc performance tests, ReadAll improves performance only when the # of file paths is very large (in the thousands and ten-thousands) -- or when taking the Union of many single-path Avro reads makes the job graph too large to submit to Dataflow.

I think it would be good to add this info about possible performance implications to the scaladoc for the `ScioContextSyntax` methods, and maybe logging a warning if `paths.size` is less than 10,000 that the user should switch to `SCollection.union(paths.map(sc.avroFile(_)))`... @regadas  / @jto wdyt?

[Beam documentation reference](https://github.com/apache/beam/blob/494256fbc1528c1ca7d0740188c1d96e23a93756/sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java#L185-L190) (note that `Read#withHintMatchesManyFiles` converts a Read to a ReadAll transform under the hood):

```java
 * <h3>Reading a very large number of files</h3>
 *
 * <p>If it is known that the filepattern will match a very large number of files (e.g. tens of
 * thousands or more), use {@link Read#withHintMatchesManyFiles} for better performance and
 * scalability. Note that it may decrease performance if the filepattern matches only a small number
 * of files.
```

Reply:
@ClaireMcGinty this is a good catch. I think logging a warning would be a good idea since small number of file patterns may match with large number of files we can't depend only on `paths.size`.

Reply:
@ClaireMcGinty updated the API doc with this information, about the warning log, I am a bit concern about adding it, since it is hard to identify the number of matching file beforehand and for the valid case, this will be polluting the output. If this is something we do in other cases in the scio then I can add.

Reply:
Yeah, agreed that it's hard to do for filepatterns. I think the documentation is 👍